dialects: add parametrised rz gate

alexarice opened the pull request:
This pr does the following:
- Adds 'quantum' dialect which contains an angle attribute which is shared by both the 'qref' and 'qssa' dialects.
- angles are stored as rational multiples of 'pi' kept to the range [0,2pi)
- Adds rz gate which is parametrised by an angle (at the moment this is given via an attribute.
- Standardises the gate set to the one used in 'http://arxiv.org/abs/1710.07345'. I can revert the removal of the cz gate if this seems like a poor idea.

~Built on top of #2819 to allow '/' in the angle syntax~
Now using ':' instead to avoid the lexing change

superlopuh reviewed (approved):
beautiful

AntonLydike commented on xdsl/dialects/quantum.py:
```diff
@@ -0,0 +1,57 @@
+    @classmethod
+    def parse_parameter(cls, parser: AttrParser) -> Fraction:
+        with parser.in_angle_brackets():
+            i = parser.parse_optional_integer()
+            numerator = 1 if i is None else i
```
I don't think an integer numerator is not covered by filecheck?

AntonLydike commented on tests/filecheck/transforms/convert_qssa_to_qref.mlir:
```diff
@@ -3,23 +3,23 @@
 // CHECK-NEXT:    qref.h %q0
-// CHECK-NEXT:    qref.cz %q1, %q0
-// CHECK-NEXT:    qref.cnot %q1, %q0
+// CHECK-NEXT:    qref.rz <pi/2> %q1
```
@PapyChacal do you know why the attribute name is not printed here?

alexarice replied:
see https://github.com/xdslproject/xdsl/issues/2807

AntonLydike reviewed (commented):
Looks good, just missing some tests covering anything that's not `<pi/2>`^^

alexarice commented on xdsl/dialects/quantum.py:
```diff
@@ -0,0 +1,57 @@
+            printer.print("pi")
+            if self.data.denominator != 1:
+                printer.print("/")
+                printer.print(self.data.denominator)
```
I've simplified this slightly

superlopuh replied:
Oh yes my bad I misread the original

n-io commented:
> * I can revert the removal of the cz gate if this seems like a poor idea.

This likely seems a higher-level, strategic consideration than could be done justice here. In my understanding there are some equivalences/semantic-preserving transformations between gates. I suppose you removed CZ in order to keep the dialect minimal? If you are wondering whether CZ should be in or not (without intending to take a position here myself), one possible consideration could be to offer a dialect with a wider variety of gates and lowering it to one with fewer gates.

alexarice commented:
Added another commit that fixed some of the arithmetic operations on angles. I also added an `AngleAttr.from_fraction` method but It would be good to know if there is another better way of doing this

alexarice commented:
Realised I would like to be able to also parse `!quantum.angle<-pi>`, working on this now

AntonLydike commented:
@alexarice what's the status on `-pi`?

alexarice commented:
I had it working but it looks like I didn't push it, I'll get it pushed over lunch

alexarice commented:
I'm very confused, it seems I managed to push the commit https://github.com/xdslproject/xdsl/commit/8e178a1e345a5590af228582109b45e0040103db , but it doesn't seem to be appearing in this pull request, even though if you click on the branch for the pull request the commit is there. Anyone know what is going on?

alexarice commented:
@AntonLydike I've rebased and force pushed and now the commit shows up, so this pr should hopefully be ready
